### PR TITLE
feat(react-native): upgrade react-native to 0.72

### PR DESCRIPTION
**Pull request by xiongemi**
## Expected Behavior

- [x] expo detox working properly
- [x] react native run and build working properly
- [x] update migration scripts to be upgraded for nx 16.6
- [x] add migration scripts to update metro.config.js
- [x] fix expo metro.config.js

## Related Issue(s)

Fixes #

### CODE REVIEW

**Comment by ste7en** [comment]
Can we have it bumped to [`0.72.1`](https://github.com/facebook/react-native/releases/tag/v0.72.1) ?

**Comment by xiongemi** [comment]
> Can we have it bumped to [`0.72.1`](https://github.com/facebook/react-native/releases/tag/v0.72.1) ?

will do. still working on it.

**Comment by xiongemi** [comment]
still have an issue with dependency conflict with @config-plugins/detox
https://github.com/expo/config-plugins/issues/178

probably after this pr got merged https://github.com/expo/config-plugins/pull/179

**Comment by xiongemi** [comment]
> still have an issue with dependency conflict with @config-plugins/detox [expo/config-plugins#178](https://github.com/expo/config-plugins/issues/178)
> 
> probably after this pr got merged [expo/config-plugins#179](https://github.com/expo/config-plugins/pull/179)

going to submit a separate pr to upgrade @config-plugins/detox in the future

File: packages/expo/src/executors/download/download.impl.ts [hunk 1]
```diff
@@ -28,6 +28,7 @@ export interface ReactNativeDownloadOutput {
 const streamPipeline = promisify(pipeline);
 
 /**
+ * @deprecated this executor is to be removed in NX 17. It is no longer used.
```
Comment (jaysoo) [inline packages/expo/src/executors/download/download.impl.ts:31]: Add a `TODO(v17)` so we can find it later.

File: packages/expo/src/generators/application/lib/add-project.ts [hunk 2]
```diff
@@ -70,6 +73,9 @@ function getTargets(options: NormalizedSchema) {
     options: {},
   };
 
+  /**
+   * @deprecated To be removed in NX 17
```
Comment (jaysoo) [inline packages/expo/src/generators/application/lib/add-project.ts:77]: Add `TODO(v17)` so we can find it later

File: packages/expo/src/migrations/update-16-6-0/update-metro-config.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * This migration updates metro.config.js to export config as a default.
+ *
+ */
+export default async function update(tree: Tree) {
+  const projects = getProjects(tree);
+
+  for (const [_, config] of projects.entries()) {
+    if (config.targets?.['start']?.executor === '@nx/expo:start') {
+      if (tree.exists(join(config.root, 'metro.config.js'))) {
```
Comment (jaysoo) [inline packages/expo/src/migrations/update-16-6-0/update-metro-config.ts:13]: What happens with their existing configuration? Should we maybe rename the old configuration and add a comment for them to port any customizations over?

File: packages/react-native/src/executors/run-android/run-android.impl.ts [hunk 4]
```diff
@@ -24,72 +19,71 @@ export default async function* runAndroidExecutor(
...
-    process.on('exit', () => childProcess.kill());
-    process.on('SIGTERM', () => childProcess.kill());
+    /**
+     * Ensure the child process is killed when the parent exits
+     */
+    const processExitListener = (signal?: number | NodeJS.Signals) => () =>
+      childProcess.kill(signal);
+    process.on('exit', processExitListener);
+    process.on('SIGTERM', processExitListener);
+    process.on('SIGINT', processExitListener);
```
Comment (jaysoo) [inline packages/react-native/src/executors/run-android/run-android.impl.ts:78]: If you listen on `SIGINT` you also need to call `process.exit()` manually, or else the process won't exit.

File: packages/react-native/src/executors/run-ios/run-ios.impl.ts [hunk 5]
```diff
@@ -28,77 +20,69 @@ export default async function* runIosExecutor(
...
-    process.on('exit', () => childProcess.kill());
-    process.on('SIGTERM', () => childProcess.kill());
+    /**
+     * Ensure the child process is killed when the parent exits
+     */
+    const processExitListener = (signal?: number | NodeJS.Signals) => () =>
+      childProcess.kill(signal);
+    process.on('exit', processExitListener);
+    process.on('SIGTERM', processExitListener);
+    process.on('SIGINT', processExitListener);
```
Comment (jaysoo) [inline packages/react-native/src/executors/run-ios/run-ios.impl.ts:77]: Same here, need to call `process.exit()` manually on `SIGINT` or else the process will not exit.

File: packages/expo/src/executors/download/download.impl.ts (hunk 1 above)
Comment (xiongemi) [reply packages/expo/src/executors/download/download.impl.ts]: added

File: packages/expo/src/generators/application/lib/add-project.ts (hunk 2 above)
Comment (xiongemi) [reply packages/expo/src/generators/application/lib/add-project.ts]: added

File: packages/expo/src/migrations/update-16-6-0/update-metro-config.ts (hunk 3 above)
Comment (xiongemi) [reply packages/expo/src/migrations/update-16-6-0/update-metro-config.ts:13]: i change this migration script. i will rename the old metro.config to metro-v71.config. i added a comment on top of the old metro config.

File: packages/react-native/src/executors/run-ios/run-ios.impl.ts (hunk 5 above)
Comment (xiongemi) [reply packages/react-native/src/executors/run-ios/run-ios.impl.ts:79]: added

**Review by jaysoo — APPROVED** [review]

**Review by juristr — APPROVED** [review]
I'd like to split this up into smaller recipes potentially, but for now this is good. We can do that in another follow-up PR later.

**Review by FrozenPandaz — APPROVED** [review]

**Pull request merged**